### ISSUE TITLE
Missed optimization to use the carry flag after subtracting

### ISSUE THREAD

**Issue by josephcsible** [llvm/llvm-project#46153]
|  |  |
| --- | --- |
| Bugzilla Link | [46809](https://llvm.org/bz46809) |
| Version | 10.0 |
| OS | Linux |
| CC | @topperc,@Laytonio,@RKSimon,@rotateright |

## Extended Description 
Consider these two C functions:

```
unsigned f1(unsigned x, unsigned y) {
    unsigned z = x - y;
    if (x < y) {
        z += 100;
    }
    return z;
}

unsigned f2(unsigned x, unsigned y) {
    unsigned z;
    if (__builtin_usub_overflow(x, y, &z)) {
        z += 100;
    }
    return z;
}
```

At -O3 or -Os, they both generate the same assembly:

        movl    %edi, %eax
        movl    %edi, %ecx
        subl    %esi, %ecx
        addl    $100, %ecx
        subl    %esi, %eax
        cmovbl  %ecx, %eax
        retq

https://godbolt.org/z/MevEM9

That's suboptimal. They should have instead used lea to save a move, save a duplicate subtraction, and avoid clobbering the carry flag, like this:

        subl    %esi, %edi
        leal    100(%rdi), %eax
        cmovael %edi, %eax
        retq

**Comment by RKSimon** [comment]
Candidate Patch: https://reviews.llvm.org/D118128

**Comment by RKSimon** [comment]
Fixed by 0b00cd19e6f3ec168b9930f235bafd5e00253dbf
```
f1:
        subl    %esi, %edi
        leal    100(%rdi), %eax
        cmovael %edi, %eax
        retq
f2:
        subl    %esi, %edi
        leal    100(%rdi), %eax
        cmovael %edi, %eax
        retq
```
